src/refiners/training_utils/batch.py

```diff
@@ -0,0 +1,222 @@
+
+        for attr_name, attr_type in attr_types.items():
+            if not attr_type in [Tensor, list]:
+                raise TypeError(f"Type of '{attr_name}' must be 'Tensor' or 'list', got '{attr_type.__name__}'")
+
+        return new_class
+
+
+class BaseBatch(metaclass=TypeCheckMeta):
+    @classmethod
+    def attr_types(cls: Type[T]) -> dict[str, Type[AttrType]]:
+        type_hints = get_type_hints(cls)
+        return {name: simple_hint(hint) for name, hint in type_hints.items()}
+
+    def __init__(self, **kwargs: AttrType):
+        attr_types = self.__class__.attr_types()
+
+        size = None
+
+        for attr_name in kwargs:
+            if attr_name not in attr_types:
+                raise ValueError(f"Attribute '{attr_name}' is not valid")
+
+        for attr_name in attr_types:
+            if attr_name not in kwargs:
+                raise ValueError(f"Missing required attribute '{attr_name}'")
+
+        for attr_name in kwargs:
+            self.__setattr__(attr_name, kwargs[attr_name], check_size=False)
+            new_size = self.attr_length(attr_name)
+            if size is not None and size != new_size:
+                raise ValueError(f"Attribute '{attr_name}' has size {new_size}, expected {size}")
+            size = new_size
+
+            if size == 0:
+                raise ValueError(f"Attribute '{attr_name}' is empty, empty attributes are not permitted")
+
+    def __getattr__(self, name: str) -> AttrType:
+        if name in self.__class__.attr_types():
+            return getattr(self, name)
+        else:
+            raise AttributeError(f"'{self.__class__.__name__}' object has no attribute '{name}'")
+
+    def __setattr__(self, name: str, value: Any, check_size: bool = True) -> None:
+        attr_types = self.__class__.attr_types()
+        if name in attr_types:
+            attr_type = attr_types[name]
+            if not isinstance(value, attr_type):
+                raise TypeError(
+                    f"Invalid type for attribute '{name}': Expected '{attr_type.__name__}', got '{type(value).__name__}'"
+                )
+            if isinstance(value, list):
+                new_size = len(value)
+            else:
+                new_size = value.shape[0]
+
+            if check_size and new_size != len(self):
+                raise ValueError(f"Attribute '{name}' has size {new_size}, expected {len(self)}")
+
+            super().__setattr__(name, value)
+        else:
+            raise AttributeError(f"'{self.__class__.__name__}' object has no attribute '{name}'")
+
+    @classmethod
+    def collate(cls: Type[T], batch_list: list[T]) -> T:
+        collated_attrs: dict[str, Any] = {}
+        attr_types = cls.attr_types()
+
+        if len(batch_list) == 0:
+            raise ValueError(f"Cannot collate an empty list of {cls.__name__}")
+
+        for attr_name, attr_type in attr_types.items():
+            attr_list = [getattr(obj, attr_name) for obj in batch_list]
+
```

Comment:
Otherwise you don't get correct typing
```suggestion
            attr_list = [obj.__getattr__(attr_name) for obj in batch_list]
```